java/datastore/src/test/java/com/google/datastore/v1/client/ChecksumEnforcingInputStreamTest.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2021 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.datastore.v1.client;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.fail;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Test for {@link ChecksumEnforcingInputStream}. */
+@RunWith(JUnit4.class)
+public class ChecksumEnforcingInputStreamTest {
+  private String payload;
+  private ChecksumEnforcingInputStream testInstance;
+  private String expectedChecksum;
+
+  public void test(int payloadSize) throws Exception {
+    // test with 10M payload
+    setUpData(payloadSize);
+    // read 1000 bytes at a time
+    // Since checksun should be correct, do not expect IOException
+    byte[] buf = new byte[1000];
+    try {
+      while (testInstance.read(buf, 0, 1000) != -1) {
+         // do nothing with the bytes read
+      }
+    } catch (IOException e) {
+      fail("checksum verification failed!");
+    }
+  }
+
+  @Test
+  public void read_withValidChecksum_differentPayloadSizes() throws Exception {
+    // test with various payload sizes (1, 2, 2**2, 2**3 etc upto 2**30 = 1GB)
```

Comment:
Some of our CI and builds run on resources constrained environments, I don't know that we need to test all the way up to 1GB every test run.

2^28 gets us up to ~268MB which I believe is likely larger than most responses and is probably a better upper bound for this test.

Reply:
thanks for the suggestion.
Done